zrml/court/src/types.rs

```diff
@@ -0,0 +1,210 @@
+    scale_info::TypeInfo,
+    Clone,
+    Debug,
+    PartialEq,
+    Eq,
+)]
+pub struct CourtInfo<BlockNumber, Appeals> {
+    /// The status of the court case.
+    pub(crate) status: CourtStatus,
+    /// The list of all appeals.
+    pub(crate) appeals: Appeals,
+    /// The information about the lifecycle of this court case.
+    pub(crate) periods: Periods<BlockNumber>,
+}
+
+impl<BlockNumber: sp_runtime::traits::Saturating + Copy, Appeals: Default>
+    CourtInfo<BlockNumber, Appeals>
+{
+    pub fn new(now: BlockNumber, periods: Periods<BlockNumber>) -> Self {
+        let pre_vote_end = now.saturating_add(periods.pre_vote_end);
+        let vote_end = pre_vote_end.saturating_add(periods.vote_end);
+        let aggregation_end = vote_end.saturating_add(periods.aggregation_end);
+        let appeal_end = aggregation_end.saturating_add(periods.appeal_end);
+        let periods = Periods { pre_vote_end, vote_end, aggregation_end, appeal_end };
+        let status = CourtStatus::Open;
+        Self { status, appeals: Default::default(), periods }
+    }
+
+    pub fn update_periods(&mut self, periods: Periods<BlockNumber>, now: BlockNumber) {
+        self.periods.pre_vote_end = now.saturating_add(periods.pre_vote_end);
+        self.periods.vote_end = self.periods.pre_vote_end.saturating_add(periods.vote_end);
+        self.periods.aggregation_end =
+            self.periods.vote_end.saturating_add(periods.aggregation_end);
+        self.periods.appeal_end = self.periods.aggregation_end.saturating_add(periods.appeal_end);
+    }
+}
+
+/// After a juror was randomly selected to vote in a court case,
+/// this information is relevant to handle the post-selection process.
+#[derive(
+    parity_scale_codec::Decode,
+    parity_scale_codec::Encode,
+    parity_scale_codec::MaxEncodedLen,
+    scale_info::TypeInfo,
+    Clone,
+    Debug,
+    PartialEq,
+    Eq,
+)]
+pub struct Draw<AccountId, Balance, Hash> {
+    /// The juror who was randomly selected.
+    pub(crate) juror: AccountId,
+    /// The weight of the juror in this court case.
+    /// The higher the weight the more voice the juror has in the final winner decision.
+    pub(crate) weight: u32,
+    /// The information about the vote state.
+    pub(crate) vote: Vote<Hash>,
+    /// The amount of funds which can be slashed for this court case.
+    /// This is related to a multiple of `MinStake` to mitigate Sybil attacks.
+    pub(crate) slashable: Balance,
+}
+
+/// All information related one item in the stake weighted juror pool.
+#[derive(
+    parity_scale_codec::Decode,
+    parity_scale_codec::Encode,
+    parity_scale_codec::MaxEncodedLen,
+    scale_info::TypeInfo,
+    Clone,
+    Debug,
+    PartialEq,
+    Eq,
+)]
+pub struct JurorPoolItem<AccountId, Balance> {
+    /// The key for the binary search to efficiently find the juror.
+    /// It has to be a unqiue key for each juror.
+    pub(crate) stake: Balance,
+    /// The account which is the juror.
```

Comment:
> sometimes it seems to refer to a vote in a court case.

What do you mean specifically?

> For example, if juror refers to accounts, then it can't really be said that the first iteration in court has 5 jurors. It might just have one juror.

Yeah it's a typical trade-off between easy explanation and the complex reality. You are referring to the juror weight and juror account id difference here. We don't really select `n` jurors out of a pool, but we select `n * MinJurorStake` out of the stake all jurors. Because it's easier to understand when saying that we select `n` jurors (although we truly select stake out of a staked pool).

Neither-less the juror account id inside a JurorPoolItem is indeed correct, because this juror account id is the basis on which we decide which account id we choose.

Reply:
c787a64e8c1a40e2268ffa1f72b915fb0ac6ef1b